src/shared/mkfs-util.c

```diff
@@ -508,19 +511,28 @@ int make_filesystem(
                 log_debug("Executing mkfs command: %s", strna(j));
         }
 
+        if (stat(node, &st) < 0 && errno != ENOENT)
+                return log_error_errno(errno, "Failed to stat %s: %m", node);
+
         r = safe_fork_full(
                         "(mkfs)",
                         stdio_fds,
                         /*except_fds=*/ NULL,
                         /*n_except_fds=*/ 0,
                         FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_DEATHSIG|FORK_LOG|FORK_WAIT|
-                        FORK_CLOSE_ALL_FDS|FORK_REARRANGE_STDIO,
+                        FORK_CLOSE_ALL_FDS|FORK_REARRANGE_STDIO|(S_ISBLK(st.st_mode) ? FORK_NEW_MOUNTNS : 0),
                         /*ret_pid=*/ NULL);
         if (r < 0)
                 return r;
         if (r == 0) {
                 /* Child */
 
+                /* mkfs.btrfs refuses to operate on block devices with mounted partitions, even if operating
+                 * on unformatted free space, so let's trick it and other mkfs tools into thinking no
+                 * partitions are mounted. */
+                if (S_ISBLK(st.st_mode) && mount_nofollow_verbose(LOG_ERR, "/dev/null", "/proc/self/mounts", NULL, MS_BIND|MS_REC, NULL) < 0)
+                        _exit(EXIT_FAILURE);
```

Comment:
i'd just do this always. and i would not fail because of this, but just do a LOG_DEBUG thing if this fails and continue

Reply:
i think it would make sense to link to the btrfs-progs issue here, so that it's clear what has to happen before we can drop this again...

Reply:
MS_REC seems a bit overkill